From 32354a84eed53f6b874619e153cd2dd27046fbc0 Mon Sep 17 00:00:00 2001
From: Mingliang Liu <liuml07@apache.org>
Date: Wed, 15 Mar 2017 10:56:02 -0700
Subject: [PATCH 2349/2848] HADOOP-14036. S3Guard: intermittent duplicate item
 keys failure. Contributed by Mingliang Liu

(cherry picked from commit b54e1b2298bf5aeef38f5db6a232199dd961aa0e)

Conflicts:
	hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/FileSystemContractBaseTest.java

Change-Id: I8ee15104a2341d8ec48aed5c12c7f3d935ff07b6
---
 .../hadoop/fs/FileSystemContractBaseTest.java      |   26 ++++++++++++++++++++
 .../org/apache/hadoop/fs/s3a/S3AFileSystem.java    |   10 ++++++--
 2 files changed, 34 insertions(+), 2 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/FileSystemContractBaseTest.java b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/FileSystemContractBaseTest.java
index 7cacc4a..52489d2 100644
--- a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/FileSystemContractBaseTest.java
+++ b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/FileSystemContractBaseTest.java
@@ -530,6 +530,32 @@ public void testOverWriteAndRead() throws Exception {
   }
 
   /**
+   * This a sanity check to make sure that any filesystem's handling of
+   * renames empty dirs doesn't cause any regressions.
+   */
+  public void testRenameEmptyToDirWithSamePrefixAllowed() throws Throwable {
+    if (!renameSupported()) return;
+    Path parentdir = path("test/parentdir");
+    fs.mkdirs(parentdir);
+    Path dest = path("test/parentdirdest");
+    rename(parentdir, dest, true, false, true);
+  }
+
+  /**
+   * This a sanity check to make sure that any filesystem's handling of
+   * renames non-empty dirs doesn't cause any regressions.
+   */
+  public void testRenameToDirWithSamePrefixAllowed() throws Throwable {
+    if (!renameSupported()) return;
+    Path parentdir = path("test/parentdir");
+    fs.mkdirs(parentdir);
+    // Before renaming, we create one file under the source parent directory
+    createFile(new Path(parentdir, "mychild"));
+    Path dest = path("test/parentdirdest");
+    rename(parentdir, dest, true, false, true);
+  }
+
+  /**
    *
    * Write a file and read it in, validating the result. Optional flags control
    * whether file overwrite operations should be enabled, and whether the
diff --git a/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java b/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java
index 806e1aa..c33bc37 100644
--- a/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java
+++ b/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java
@@ -924,8 +924,14 @@ private boolean innerRename(Path source, Path dest)
         }
       }
 
-      // We moved all the children, now move the top-level dir.
-      S3Guard.addMoveDir(metadataStore, srcPaths, dstMetas, src, dst, username);
+      // We moved all the children, now move the top-level dir if it's non-empty
+      // Empty directory has been added as the object summary of listObjects
+      if (srcStatus.isEmptyDirectory() == Tristate.FALSE) {
+        LOG.debug("To move the non-empty top-level dir src={} and dst={}",
+            src, dst);
+        S3Guard.addMoveDir(metadataStore, srcPaths, dstMetas, src, dst,
+            username);
+      }
     }
 
     metadataStore.move(srcPaths, dstMetas);
-- 
1.7.9.5

